modules/systemdunits/collect.go

```diff
@@ -0,0 +1,94 @@
+package systemdunits
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+)
+
+func convertUnitState(state string) int64 {
+	switch state {
+	case "active":
+		return 1
+	case "activating":
+		return 2
+	case "failed":
+		return 3
+	case "inactive":
+		return 4
+	case "deactivating":
+		return 5
+	default:
+		return -1
+	}
+
+}
+
+func (s *SystemdUnits) collect() (map[string]int64, error) {
+	conn, err := dbus.New()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	allUnits, err := conn.ListUnits()
+	if err != nil {
+		return nil, err
+	}
+
+	mx := make(map[string]int64)
+
+	units := s.filterUnits(allUnits)
+	for _, unit := range units {
+		ut, err := extractUnitType(unit.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		chartID := fmt.Sprintf("%s_states", ut)
+		chart := s.charts.Get(chartID)
+		if !chart.HasDim(unit.Name) {
+			_ = chart.AddDim(&Dim{ID: unit.Name})
+		}
```

Comment:
`charts` and `dimensions` are slices, `Get` and `HasDim` methods are not fast - it is iterating over the slice.

This is a problem and should be fixed, but for now you need to keep track of `collected` services in the module.

See example:
https://github.com/netdata/go.d.plugin/blob/79ffe8973780b29f1edae13550e0070fdfb69c37/modules/filecheck/collect_files.go#L18-L22

Reply:
I got this idea from other modules, like this one
https://github.com/netdata/go.d.plugin/blob/79ffe8973780b29f1edae13550e0070fdfb69c37/modules/k8s_kubelet/collect.go#L46-L49

Reply:
i see, it is not good, imagine iterating over the same slice on every iteration number of units times again and again.